include/oneapi/dpl/pstl/utils.h

```diff
@@ -669,6 +559,151 @@ __dpl_ceiling_div(_T1 __number, _T2 __divisor)
+        --__my_cn;
+        return *this;
+    }
+
+    template <typename _DiffType>
+    _ReverseCounter&
+    operator+=(_DiffType __val)
+    {
+        __my_cn -= __val;
+        return *this;
+    }
+
+    difference_type
+    operator-(const _ReverseCounter& __a)
+    {
+        return __a.__my_cn - __my_cn;
+    }
+
+    operator _IntType() { return __my_cn; }
+
+// TODO: Temporary hotfix. Investigate the necessity of _ReverseCounter
+// Investigate potential user types convertible to integral
+// This is the compile-time trick where we define the conversion operator to sycl::id
+// conditionally. If we can call accessor::operator[] with the type that converts to the
+// same integral type as _ReverseCounter (it means that we can call accessor::operator[]
+// with the _ReverseCounter itself) then we don't need conversion operator to sycl::id.
+// Otherwise, we define conversion operator to sycl::id.
+#if _ONEDPL_BACKEND_SYCL
+    struct __integral
+    {
+        operator _IntType();
+    };
+
+    template <typename _Tp>
+    static auto
+    __check_braces(int) -> decltype(::std::declval<_Tp>()[::std::declval<__integral>()], ::std::false_type{});
+
+    template <typename _Tp>
+    static auto
+    __check_braces(...) -> ::std::true_type;
+
+    class __private_class;
+
+    operator ::std::conditional_t<decltype(__check_braces<_Acc>(0))::value, sycl::id<1>, __private_class>()
+    {
+        return sycl::id<1>(__my_cn);
+    }
+#endif
+};
+
+// Reverse searching for the first element strongly less than a passed value - left bound
+template <typename _Buffer, typename _Index, typename _Value, typename _Compare>
+_Index
+__pstl_left_bound(_Buffer& __a, _Index __first, _Index __last, const _Value& __val, _Compare __comp)
+{
+    auto __beg = _ReverseCounter<_Index, _Buffer>{__last - 1};
+    auto __end = _ReverseCounter<_Index, _Buffer>{__first - 1};
+
+    return __pstl_upper_bound(__a, __beg, __end, __val, __reorder_pred<_Compare>{__comp});
+}
+
+// Lower bound implementation based on Shar's algorithm for binary search.
+template <typename _Acc, typename _Size, typename _Value, typename _Compare>
+_Size
+__shars_lower_bound(_Acc __acc, _Size __first, _Size __last, const _Value& __value, _Compare __comp)
+{
+    static_assert(::std::is_unsigned_v<_Size>, "__shars_lower_bound requires an unsigned size type");
+    const _Size __n = __last - __first;
+    if (__n == 0)
+        return __first;
+    _Size __cur_pow2 = __dpl_bit_floor(__n);
+    const _Size __midpoint = __n / 2;
+    // Check the middle element to determine if we should search the first or last
+    // 2^(bit_floor(__n)) - 1 elements.
+    const _Size __shifted_first = __comp(__acc[__midpoint], __value) ? __n + 1 - __cur_pow2 : __first;
+    // Check descending powers of two. If __comp(__acc[__search_idx], __pow) holds for a __cur_pow2, then its
+    // bit must be set in the result.
+    _Size __search_offset{0};
+    for (__cur_pow2 /= 2; __cur_pow2 > 0; __cur_pow2 /= 2)
```

Comment:
My assumption is that the compiler will always replace these divisions by two with shift right by one.  You've probably checked performance comparison, if not it could be worth a look.

So if they are the same performance, then we have to decide which is more clear to read.  My preference for the ones in this loop would be to use `>> 1`, since we are effectively setting each bit of the index with each loop iteration, and we use bitwise ops elsewhere in the loop.

Just my preference, feel free to leave as is.

Reply:
I read that `operator>>' isn't so clear as we thinking for the signed types with the value `< 0`. The interesting moment is the value of sign bit and highest bit(s) which state should be assigned during the `operator>>' implementation. See https://stackoverflow.com/questions/141525/what-are-bitwise-shift-bit-shift-operators-and-how-do-they-work for details. So would prefer `/2` operation when the type is unknown.

Reply:
I agree with you for unknown types, but this function has a static_assert enforcing that `_Size` is an unsigned type.

Reply:
So it's OK in this case.

Reply:
Performance is identical between these two versions, and unsigned types are required for the bit floor call.

I wasn't sure which one is clearer when writing this, but I think `>>=` makes the most sense since we use other bitwise operations in the loop.

I have went ahead and made this change.